### ISSUE TITLE
Cannot index record with own key type

### ISSUE THREAD

**Issue by Nixinova** [microsoft/TypeScript#45159]
# Bug Report

### 🔎 Search Terms

* error = `Element implicitly has an 'any' type because expression of type X can't be used to index type 'Record<X, Y>'`
* template literal tag type invalid in key

### 🕗 Version & Regression Information

- [ ] This is a crash
- [x] This changed between versions 4.0.x and 4.1.x
- [ ] This is the behavior in every version I tried, and I reviewed the FAQ for entries
- [ ] I was unable to test this on prior versions

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBAahBOUC8UAGASA3gZ2PAlgHYDmAvgHRa4EmmoBQ9AxgPaG6wLb5vYBcUAEoRW8ACYAeOPAA0UakWIA+ZFEz0omqAHIAjOQAM2gdoBmLFtpkate8ruM6ARgEN4V+qUYAbCMCgAbggC0qp2RvTS3LwA2kHwALr0APTJWukAelAAovDwLPAC2b4AthCE-vglYN74TPjA3iBQABYu2FAuhDpdINpQoJBQTiIuAK7Y0BAAHmDwENjR3SymKWnp6YPQ2hg4eIoUVPu0qP1MXdr+I1ATEGIDLFBEYjMD4BBrG+nawqKSuwpaJQ9jQyKg5IDlNpyFAABQAdgMAFYAMwASiAA)

### 💻 Code

```ts
type Ver = `${string}.${string}`

const data: Record<Ver, string> = {
    '1.0': 'foo',
    '1.1': 'bar',
}

let ver: Ver = '1.0'
data[ver] // Error 7053
```

### 🙁 Actual behavior

``Versions[ver]`` errors with ``Element implicitly has an 'any' type because expression of type '`${string}.${string}`' can't be used to index type 'Record<`${string}.${string}`, string>'.`` (??)

### 🙂 Expected behavior

TypeScript recognises that `` `${string}.${string}` `` === the key of ``Record<`${string}.${string}`,string>``

**Comment by Nixinova** [comment]
Seems to be a duplicate of #40193 which in turn is marked as a duplicate of #37448 though I don't really understand the latter as the issue is philosophically worded.

Why doesn't my code work?; I'll close when that's answered.

**Comment by ahejlsberg** [comment]
Duplicate of #42192. Fixed in 4.4-beta and later.

**Comment by Nixinova** [comment]
Oh, okay, thanks.
